### PR TITLE
subq/join

### PR DESCRIPTION
**Pull request by earonesty**

### CODE REVIEW

File: tests/test_notanorm.py [hunk 1]
```diff
@@ -153,6 +153,25 @@ def __init__(self, bar=None):
             pass
 
 
+def test_db_select_gen_close_fetch(db):
```
Comment (colatkinson) [inline tests/test_notanorm.py:156]: What's this testing?

**Review by colatkinson — APPROVED** [review]

File: notanorm/base.py [hunk 2]
```diff
@@ -93,10 +105,223 @@ def __eq__(self, o: Any) -> bool:
         return self.op == o.op and self.val == o.val
 
 
-class SubQ:
-    def __init__(self, sql: str, vals: Tuple[Any] = ()):
+class BaseQ(ABC):
+    _next = 0
+    fields = []
+    field_map = {}
```
[the remark this answers is not in the document]
Comment (earonesty) [reply notanorm/base.py:111]: ill try that, yeah, it was a pycharm thing

File: notanorm/base.py [hunk 3]
```diff
@@ -93,10 +105,223 @@ def __eq__(self, o: Any) -> bool:
...
+
+    def __init__(self, db: "DbBase"):
+        self.db = db
+
+    @classmethod
+    def next_num(cls):
+        cls._next = cls._next + 1 % 100000
```
[the remark this answers is not in the document]
Comment (earonesty) [reply notanorm/base.py:118]: 1. so it doesn't get too big
2. its only every used inside a lock, so probably dont have to worry about that

File: tests/test_notanorm.py (hunk 1 above)
Comment (earonesty) [reply tests/test_notanorm.py:156]: that the generator closes on fetch

File: notanorm/base.py [hunk 4]
```diff
@@ -922,3 +1263,45 @@ def select_one(self, table, fields=None, **where):
...
+        """
+        ret = self.select_gen(table, fields, **where)
+        try:
+            return next(ret)
+        except StopIteration:
+            return None
+
+    def _get_table_cols(self, tab: str):
+        if not self.__model_cache:
+            self.__model_cache = self.model()
```
[the remark this answers is not in the document]
Comment (earonesty) [reply notanorm/base.py:1282]: passing a model in should be fine

**Pull request merged**